Implement AsyncNextMessageDirect (fixes #150)

MaxMaeder opened the pull request:
I've implemented AsyncNextMessageDirect and FrameAssembler as defined in #150, plus tests for both. 

The one issue I ran into is that the websocket stream and codec are both designed to only store one frame in memory at a time. I tried to rewrite this so we could buffer an entire message's worth of frames (so it would be zero-copy), but I kept running into issues as this was a pretty complicated refactor.

The solution I came up with was:
1) If it's a single frame message, just return the reference to the payload (this is zero copy)
2) If the message is multiple frames, copy the intermediate frames to an internal buffer

This obviously is not a perfect solution. I would be happy to continue trying to implement a true zero-copy method, but I would probably need some help or guidance.

sergiu128 commented on codec/websocket/stream.go:
```diff
@@ -1079,3 +1079,134 @@ func (s *Stream) CloseNextLayer() (err error) {
+//   - an error occurs when reading/decoding the message bytes from the underlying stream
+//   - the payload of the message is successfully read
+func (s *Stream) AsyncNextMessageDirect(callback AsyncMessageDirectCallback) {
+	s.AsyncFlush(func(err error) {
```
I don't think this is needed as `AsyncNextFrame` flushes already before reading the next frame. Am I missing something here?

sergiu128 commented on codec/websocket/stream.go:
```diff
@@ -1079,3 +1079,134 @@ func (s *Stream) CloseNextLayer() (err error) {
+
+// asyncNextMessageDirectFirstFrame reads the first data frame.
+// If it is FIN, we do single-frame zero-copy. Otherwise, we start gathering frames.
+func (s *Stream) asyncNextMessageDirectFirstFrame(cb AsyncMessageDirectCallback) {
```
The first idea is good! I think here we can reduce the amount of code to around half.

When dealing with fragmented reads/writes or in general with any operation that involves multiple underlying calls, try to always think what is the information that you need on each underlying call to complete the full operation. In some cases the information is non-existent, which makes the call stateless. In other cases, the information is at least some set of variables, which makes the call stateful.

For example, `AsyncNextMessage` is a stateful operation - its state is baked into its arguments: `continuation, readBytes, b`. Another example are the reads/writes in `async_adapter`. [There](https://github.com/talostrading/sonic/blob/master/async_adapter.go#L33) you introduced the state as `struct` which is a bit more explicit. 

`AsyncNextMessageDirect` is a stateful operation. The state can be explicit (a `struct`, like in `async_adapter`) and is most likely comprised of:
- `[][]parts` or `[][]payloads` which gets updated on each call - it's resulting length will be at least 1
- `totalSoFar` which is the rolling sum of all the read payloads so far. We need this to ensure we respect the max message size, as you've already done. Putting this one into the operation's "state" saves having to iterate over all read parts in `asyncNextMessageDirectFragments`. Instead, you can updated it on each call, so "state change" and then check that it's within bounds.

By introducing an explicit state, you'll be able to collapse `asyncNextMessageDirectFragments` and `asyncNextMessageDirectFirstFrame` into a single function.

sergiu128 commented on codec/websocket/stream.go:
```diff
@@ -1079,3 +1079,134 @@ func (s *Stream) CloseNextLayer() (err error) {
+		}
+
+		// If message is made of a multiple frames, we must copy each frame's payload
+		// This is because existing codec/ws stream designed around only holding on to one frame at a time
```
Nicely noted here: having the codec operate on a single frame allocated at init time prevents us from making truly "zero-copy" reads.

I think the only option we have here is to allow the `FrameCodec` to hold multiple frames at once. The codec connection in `codec.go` only deals with `Encode` and `Decode` so we're free to do whatever we want with the codec itself. In the end, we need to have a way to "modify" the codec while reading messages in `stream.go`. This modification can be part of the `AsyncNextMessageDirect`'s state update. I'm thinking of the following, although this might not be the best:
- introduce a method `ReserveFrame` in codec that ensures the currently read frame won't be overwritten in the next decode call. This method will allocate a new frame to decode into, either with `NewFrame` or with some custom decoder
  - one idea is to have a single `[]byte` slice for all frames, but this is an optimization is for later. We can size this to be at least `maxMessageSize` + the size of say `10` frame headers.
- we call `ReserveFrame` after each read fragment that's a continuation
- after we read the last frame, we invoke the user's callback (passing it each frame's payloads), and then we can call `codec.ReleaseFrames` to force the codec to reset its state before the next non-direct/direct read.

I think we should first focus on making the mechanism work i.e. create a correct interaction between `stream.go` and `frame.go` that allows us to store multiple frames throughout a read, as outline above. This is just one solution, there's probably something better so you'd need to tinker a bit here.

After that we can look on how to best keep multiple frames in `codec.go`. We don't want to allocate on each fragmented direct read, and it would be ideal if all the frames are close to one another in memory.

MaxMaeder commented:
Hi Sergiu,

First, thanks for you patience as I learn more about how Sonic works.

I've made some changes following from your review, let me know what you think.

First, I modified `frame_codec.go` to add `ReserveFrame`/`ReleaseFrames`. However I implemented this slightly differently to how you indicated. If you look in the frame codec, `decodeFrame` is initialized with an allocation of a frame using `NewFrame()`. But the codec immediately replaces this allocation by making `decodeFrame` reference memory in the underlying `ByteBuffer` when doing a decode. I think then that this allocation was originally done just to avoid nil-pointer edge cases if methods in the codec were called before a decode.

I'm bringing this up because my interpretation of your idea with `ReserveFrame` was to essentially swap out the region of memory we're going to read the next frame into and therefore be able to store multiple frames at once. Since `FrameCodec` *doesn't actually own any memory*, this wouldn't work. **My understanding is that FrameCodec.Decode() just figures out what slice of ByteBuffer's we need to hold a reference to the next complete frame.**

My idea then was to just allocate a messageBuffer in `FrameCodec` of `maxMessageSize` bytes. Then, `ReserveFrame` will just append the message payload from the latest decoded frame (which is just a slice of the `ByteBuffer`) into this buffer. Then, we can simply add a method `MessagePayload` which returns a reference to this buffer containing the message payload. `ReleaseFrames` just resets this buffer for the next message.

With this, our logic in `stream.go` becomes much simpler, and although we are doing a memcpy for all the payload components, we are not doing any expensive dynamic allocation and we can avoid something like `FrameAssembler` altogether.

Let me know what you think of this solution, or if I am still misunderstanding how the WebSocket implementation works. I've also still left in `FrameAssembler` for the time being, if this solution is inadequate.

sergiu128 commented:
Thank you for your changes! Sorry for slow replies, had a lot of work to tie up as I'm off for a couple of days. I'll reply faster from now on :)

This change is good, I think we're going in the right direction. `stream.go` is indeed much simpler here. And you observed right: the frame codec just figures out what part of the ByteBuffer we need to hold a reference to. This fully slipped my mind while I was typing out my ideas above, long time since I've dealt with ByteBuffer, sorry! The line that gives it away is `c.decodeFrame = src.Data()[:readSoFar]` in [here](https://github.com/talostrading/sonic/blob/master/codec/websocket/frame_codec.go#L59).

I still think we should explore getting rid of memory copies all together in `stream.go` in this new read function. You are right that dynamic allocations are on average more costly, but memory copies can also blow out the same way. Usually the cost of a memory copy is linear to the amount of bytes copied, so we'd get bit on big messages.

Maybe we can use the `ByteBuffer` to our advantage. There is this method [Save](https://github.com/talostrading/sonic/blob/master/byte_buffer.go#L164) which, if invoked, will keep the bytes from the read area for later consumption. I think we will still be able to use the `PrepareRead/Commit/Consume` semantics without disturbing any saved bytes in this codec.

By doing this we isolate the memory copies purely to the ByteBuffer. So we'd cut down from a memory copy in stream.go + a memory copy in byte buffer (to clear the read message through either `Consumer` or `Discard`), to just a memory copy in the byte buffer. The next little project will then be to make the byte buffer zero-copy (there's a start on that [here](https://github.com/talostrading/sonic/blob/master/bytes/mirrored_buffer.go)).

I'd start with playing with the ByteBuffer. I think it's pretty well [documented](https://github.com/talostrading/sonic/blob/master/byte_buffer.go#L9). See how we can use this `Save` function and how we can adapt the codec to make use of that (and also how to reference the saved frames, probably through some sort of `ReserveFrame` function).

I won't have any work for the next couple of days so will reply promptly, and also have time to code on sonic a bit. Feel free to post any findings or questions here, no need to wait a lot for a good/perfect solution.

MaxMaeder commented:
Thanks for bearing with me while I learn how all of the components of the library work. 

ByteBuffer is super nice for tasks like this! I rewrote how `AsyncNextMessageDirect` works to take advantage of the save area, let me know what you think!

sergiu128 commented:
Great! Was traveling the past few days, taking a look tomorrow!

sergiu128 commented on codec/websocket/stream.go:
```diff
@@ -1079,3 +1079,78 @@ func (s *Stream) CloseNextLayer() (err error) {
 	return
 }
+
+// AsyncNextMessageDirect reads the next websocket message asynchronously, providing a buffer
```
nitpick: here it should be smth like: "providing one more slices representing each payload fragment of a message. If the next message is unfragmented, then a single byte slice is provided, representing the message's full payload. If the message is fragmented into `n` fragments, then `n` payloads are provided. Callers can reconstruct the full message payload by a copying out the provided slices into a continuous memory chunk. "

sergiu128 commented on codec/websocket/frame_codec.go:
```diff
@@ -110,3 +116,39 @@ func (c *FrameCodec) Encode(frame Frame, dst *sonic.ByteBuffer) error {
+}
+
+// ReleaseFrames releases our reserved frames
+func (c *FrameCodec) ReleaseFrames() {
```
I think you are correct in that you don't need `decodeReset = true` in this function but I would make a note of it. 
The `DiscardAll` ensures proper cleanup after reading a fragmented message.

Do note though that there's now a difference in lifetimes of the frames between the direct and non direct versions. For the non-direct case, a frame is valid until the next `AsyncNextMesssage` invocation (which calls `resetDecode`). For the direct case in this PR, frames are valid only within the `AsyncNextMessageDirect` callback and not after, as we call `DiscardAll` immediately after the callback. That means we can't store references to the frame payloads after `AsyncNextMessageDirect` returns, and before the next one is invoked.

Do you think we can prolong the lifetime of the frames in the direct case to be valid between two `AsyncNextMessageDirect` calls and not only within the callback? Not a huge requirement especially if it complicates the code, but would be nice to have :) . I think the tricky case will be to support interleaved control messages here.

sergiu128 replied:
I think this is the simplest solution, I like it!

sergiu128 replied:
Would it also make sense to call `ReleaseFrames` in `AsyncNextMessage`? Unsure, what do you think?

MaxMaeder replied:
I think it would be simplest to just release in `AsyncNextMessageDirect`, since the save area isn't used anywhere else and it keeps the logic structurally close together.

sergiu128 reviewed (changes requested):
Very clean and neat solution, I really like it! Thank you! I think we're very close to a prod-ready solution here. I left some inline comments. Besides those, I think we need a couple more tests:
- test with max message size breached by first fragment
- test with max message size breached by a fragment following the first one
- test with nth fragment where n > 1 without continuation set, to ensure `ErrExpectedContinuation` propagates correctly
- test with multiple `AsyncNextMessageDirect` done in sequence
- test with `AsyncNextMessageDirect`, `AsyncNextMessage` (so normal, non zero-copy), `AsyncNextMessageDirect` and so on..
- test with the [websocket client validator](https://github.com/talostrading/sonic/tree/master/tests/autobahn)
  - install docker make sure it's running
  - call `./run.sh` in a terminal
  - modify `client.go` to use `AsyncNextMessageDirect`
  - call `go run client.go`
  - the tests results should be generated in some path of the `autobahn` folder, can see them by just opening `index.html`

sergiu128 commented on codec/websocket/frame_assembler.go:
```diff
@@ -0,0 +1,38 @@
+}
+
+// Reassemble concatenates all slices into a single new []byte buffer.
+func (fa *FrameAssembler) Reassemble() []byte {
```
I would also add a function `ReassembleInto(b []byte)` here, so callers can avoid any allocations if they want to.

sergiu128 commented on codec/websocket/frame_assembler.go:
```diff
@@ -0,0 +1,47 @@
+func (fa *FrameAssembler) ReassembleInto(b []byte) {
+	offset := 0
+	for _, p := range fa.parts {
+		copy(b[offset:], p)
```
this might panic if b cannot fit the current part. In this case we can just return `false` from this function. If the reassembly is successful, we can return `true`. Odds are the callers will already size `b` to the max messsage size configured for the stream, but better be safe :)

sergiu128 commented on codec/websocket/frame_codec.go:
```diff
@@ -110,3 +116,39 @@ func (c *FrameCodec) Encode(frame Frame, dst *sonic.ByteBuffer) error {
+
+// ReservedFramePayloads returns references to the payloads of our reserved frames
+func (c *FrameCodec) ReservedFramePayloads() [][]byte {
+	return c.messagePayloads;
```
nice optimization here

sergiu128 commented:
> I think I addressed the implementation changes you requested, can you double check before I move on to writing the additional tests?
> 
> Thanks, Max

Changes look good, can move on with testing, thank you! For the test where we combine `AsyncNextMessageDirect` and `AsyncNextMessage`, we should also assert that the `ByteBuffer` save and read areas are in order. This depends on what you come up with [here](https://github.com/talostrading/sonic/pull/153#discussion_r1945197546). Essentially we just want to make sure that all data in the buffer eventually gets flushed out. So nothing stays in forever. I think in the current solution you can have the read and save area occupied at the same time which is fine. We just want to make sure they don't grow indefinitely.

sergiu128 commented on tests/autobahn/client.go:
```diff
@@ -103,14 +103,17 @@ func runTest(i int) {
 		b := make([]byte, 1024*1024)
 
-		var onAsyncRead websocket.AsyncMessageCallback
+		var onAsyncRead websocket.AsyncMessageDirectCallback
```
can we have both versions here? So through we flag we can choose whether to test AsyncNextMessage or AsyncNextMessageDirect...

sergiu128 commented on codec/websocket/stream.go:
```diff
@@ -1124,6 +1133,14 @@ func (s *Stream) asyncNextMessageDirect(
 			// If this is the first frame we decode, get the message type from the opcode
 			messageType = MessageType(f.Opcode())
+
+			if f.Opcode().IsContinuation() {
```
these were reported by autobahn? Nice!

MaxMaeder replied:
Yes, it's a cool tool!

sergiu128 commented on examples/derabit/main.go:
```diff
@@ -0,0 +1,107 @@
+		n, _ := ioc.PollOne()
+		if n > 0 {
+			eventsReceived += n
+			ioLatency.Add(float64(time.Since(start).Microseconds()))
```
do you see a difference compared to the non-direct/memcpy version? On my mac it's around 10 to 50 micros.

sergiu128 replied:
<img width="606" alt="image" src="https://github.com/user-attachments/assets/4f431faa-f6dc-4880-b198-7ded7fcc9073" />

reassemble into still copies here, so there's not much gain for single frame messages compared to the non-direct version. Seems that deribit always sends 1 frame messages. I think here we just want to access `pl[0]` for single frame messages and only use the frame assembler for multi frame messages.

MaxMaeder replied:
Ok, I've removed use of the `FrameAssembler` for Derabit.

MaxMaeder commented:
I've made the requested changes, let me know what you think!

sergiu128 commented on examples/derabit/main.go:
```diff
@@ -0,0 +1,103 @@
+package main
```
should be renamed to "deribit" from "derabit"

sergiu128 commented on examples/derabit/main.go:
```diff
@@ -0,0 +1,103 @@
+				}
+
+				if len(pl) != 1 {
+					panic("Derabit should only be sending single frame messages.")
```
deribit*

sergiu128 reviewed (changes requested):
All looks good, thank you for implementing the changes and all the work. 3 things left:
- derabit should be deribit
- `TestClientMixedAsyncNextMessageDirectAndAsyncNextMessage` is failing for me, i think `buf` needs to be sliced to a smaller length cuz the payloads look good but i'll let you take a look
- pls rebase with master, I introduced some changes around setting the `MaxMessageSize` for a stream pre-handshake

MaxMaeder commented:
Should all be fixed!
